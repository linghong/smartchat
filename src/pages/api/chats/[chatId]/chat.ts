import 'reflect-metadata';
import { NextApiRequest, NextApiResponse } from 'next';
import {
  DataSource,
  EntityNotFoundError,
  ConnectionIsNotSetError,
  EntityMetadataNotFoundError
} from 'typeorm';

import { getAppDataSource, Chat, User, ChatMessage, ChatImage } from '@/src/db';

const handleDeleteRequest = async (
  res: NextApiResponse,
  dataSource: DataSource,
  chatIdNum: number
) => {
  // Start a transaction
  await dataSource.transaction(async transactionalEntityManager => {
    // Find the chat
    const chat = await transactionalEntityManager.findOne(Chat, {
      where: { id: chatIdNum },
      relations: ['messages', 'messages.images']
    });

    if (!chat) {
      return res.status(404).json({ message: 'Chat not found' });
    }

    // Delete all images associated with the chat's messages
    if (chat.messages) {
      for (const message of chat.messages) {
        if (message.images) {
          await transactionalEntityManager.delete(ChatImage, {
            messageId: message.id
          });
        }
      }
    }

    // Delete all messages associated with the chat
    await transactionalEntityManager.delete(ChatMessage, { chatId: chat.id });

    // Finally, delete the chat itself
    await transactionalEntityManager.delete(Chat, { id: chat.id });
  });

  res
    .status(200)
    .json({ message: 'Chat and all associated data deleted successfully' });
};

const handlePutRequest = async (
  req: NextApiRequest,
  res: NextApiResponse,
  dataSource: DataSource,
  chatIdNum: number
) => {
  const { title } = req.body;
  const chatRepository = dataSource.getRepository(Chat);

  if (!title) {
    return res.status(400).json({ message: 'Title is required' });
  }

  if (typeof title !== 'string' || title.length < 1 || title.length > 255) {
    return res
      .status(400)
      .json({ message: 'Title must be a string between 1 and 255 characters' });
  }

  // First, find the chat
  const chat = await chatRepository.findOne({ where: { id: chatIdNum } });

  if (!chat) {
    return res.status(404).json({ message: 'Chat not found' });
  }

  // Update the chat title
  chat.title = title;
  await chatRepository.save(chat);

  res.status(200).json({ message: 'Chat title updated successfully' });
};

const handler = async (req: NextApiRequest, res: NextApiResponse) => {
  // validation
  const { chatId } = req.query;
  const chatIdNum = Number(chatId);
  if (isNaN(chatIdNum)) {
    return res.status(400).json({ message: 'Invalid chat ID' });
  }

  try {
    const dataSource = await getAppDataSource();

    if (!dataSource)
      return res.status(500).json({ message: 'AppDataSource is null' });

    switch (req.method) {
      case 'DELETE':
        await handleDeleteRequest(res, dataSource, chatIdNum);
        break;
      case 'PUT':
        await handlePutRequest(req, res, dataSource, chatIdNum);
        break;
      default:
        res.setHeader('Allow', ['DELETE', 'PUT']);
        res.status(405).end(`Method ${req.method} Not Allowed`);
    }
  } catch (error) {
    console.error('Error during Chat operation', error);
    if (error instanceof EntityNotFoundError) {
      res.status(404).json({ message: 'Chat not found' });
    } else if (error instanceof ConnectionIsNotSetError) {
      res.status(500).json({ message: 'Database connection error' });
    } else if (error instanceof EntityMetadataNotFoundError) {
      res.status(500).json({ message: 'Entity metadata error' });
    } else {
      res.status(500).json({ message: 'Internal server error' });
    }
  }
};

export default handler;
