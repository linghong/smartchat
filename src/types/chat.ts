export interface Chat {
  id: number;
  title: string;
}

export type ChatRole = 'system' | 'assistant' | 'user';

export type ChatType = {
  role: ChatRole;
  content: string;
};

export interface ImageFile {
  base64Image: string;
  mimeType: string;
  size: number;
  name: string;
  width?: number;
  height?: number;
}

export interface FileData {
  base64Content: string;
  type: string;
  size: number;
  name: string;
}

export interface Message {
  question: string;
  answer: string;
  model: string;
}

export interface OpenAIChatContentImage {
  type: 'image_url';
  image_url: {
    url: string;
    detail: 'auto' | 'low' | 'high';
  };
}
export interface OpenAIChatContentText {
  type: 'text';
  text: string;
}
