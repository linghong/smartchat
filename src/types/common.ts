export interface OptionType {
  value: string;
  label: string;
  tags?: string[];
  category?: string;
  contextWindow?: number;
  vision?: boolean;
}

export interface ChatOption extends OptionType {
  tags: string[];
}

export type ApiResponse = {
  message: string;
  fileName: string;
  error?: string;
};

export type InputErrors = {
  [key: string]: string | null;
};

export type Input<T> = {
  [key: string]: T;
};

export type InputData = {
  [key: string]: string | number | null;
};

export type UploadData = {
  [key: string]: File | null;
};

export type UploadErrors = {
  [key: string]: string | null;
};
