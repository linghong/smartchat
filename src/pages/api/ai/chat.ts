import type { NextApiRequest, NextApiResponse } from 'next';
import { marked } from 'marked';

import { fetchDataFromPinecone } from '@/src/services/fetchDataFromPinecone';

import getClaudeChatCompletion from '@/src/services/llm/claude';
import getGeminiChatCompletion from '@/src/services/llm/gemini';
import { getGroqChatCompletion } from '@/src/services/llm/groq';
import {
  createEmbedding,
  getOpenAIChatCompletion
} from '@/src/services/llm/openai';
import getOpenModelChatCompletion from '@/src/services/llm/opensourceai';

import {
  processImageFiles,
  processNonMediaFiles
} from '@/src/utils/processMessageFile';
import { FileData, ImageFile } from '@/src/types/chat';
import {
  extractMessageContent,
  extractSubjectTitle
} from '@/src/utils/chatMessageHelper';

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  const { basePrompt, namespace, selectedModel, chatHistory, fileSrc } =
    req.body;

  let { question } = req.body;

  if (!question && fileSrc.length === 0) {
    return res.status(400).json({ message: 'No question in the request' });
  }

  if (!selectedModel?.value) {
    return res.status(500).json('Model name is missing.');
  }

  const base64ImageSrc = processImageFiles(fileSrc);

  question += await processNonMediaFiles(fileSrc);

  // Replace multiple newlines with a single newline to avoid excessive spacing
  const pretreatedQuestion = question.trim().replace(/\n\s*\n/g, '\n');

  try {
    let fetchedText = '';

    //fetch text from vector database
    if (namespace && namespace !== 'none') {
      const embeddedQuery = await createEmbedding(question);
      fetchedText = await fetchDataFromPinecone(embeddedQuery, namespace);
    }

    // get response from AI
    const { category } = selectedModel;
    let chatResponse: string | undefined | null;

    switch (category) {
      case 'openai':
        chatResponse = await getOpenAIChatCompletion(
          basePrompt,
          chatHistory,
          pretreatedQuestion,
          fetchedText,
          selectedModel,
          base64ImageSrc
        );
        break;

      case 'groq':
        chatResponse = await getGroqChatCompletion(
          basePrompt,
          chatHistory,
          pretreatedQuestion,
          fetchedText,
          selectedModel
        );
        break;

      case 'google':
        chatResponse = await getGeminiChatCompletion(
          basePrompt,
          chatHistory,
          pretreatedQuestion,
          fetchedText,
          selectedModel,
          base64ImageSrc
        );
        break;

      case 'anthropic':
        chatResponse = await getClaudeChatCompletion(
          basePrompt,
          chatHistory,
          pretreatedQuestion,
          fetchedText,
          selectedModel,
          base64ImageSrc
        );
        break;

      case 'hf-small':
      case 'hf-large':
        const baseUrl =
          category === 'hf-small'
            ? process.env.NEXT_PUBLIC_SERVER_URL
            : process.env.NEXT_PUBLIC_SERVER_GPU_URL;
        if (!baseUrl) {
          return res
            .status(500)
            .json(
              `Url address for posting the data to ${selectedModel} is missing`
            );
        }
        const url = `${baseUrl}/api/chat_${category === 'hf-small' ? 'cpu' : 'gpu'}`;
        chatResponse = await getOpenModelChatCompletion(
          basePrompt,
          chatHistory,
          pretreatedQuestion,
          fetchedText,
          selectedModel,
          url
        );
        break;

      default:
        return res.status(500).json('Invalid model category');
    }

    let answer: string;
    let subject: string;
    if (!chatResponse) {
      answer = "Sorry, I'm having trouble finding an answer to your question.";
      subject = 'Unknown';
    } else {
      answer = extractMessageContent(chatResponse);
      subject = extractSubjectTitle(chatResponse);
    }

    const htmlAnswer = marked(answer);

    res.status(200).json({ answer: htmlAnswer, subject });
  } catch (error: any) {
    console.error('An error occurred: ', error);
    res.status(500).json({ error: error.message || 'Something went wrong' });
  }
}
