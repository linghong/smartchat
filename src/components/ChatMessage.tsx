import Image from 'next/image'

import { Message, ImageFile } from '@/src/types/chat'
import ImageListWithModal from '@/src/components/ImageListWithModal'

type ChatMessageProps = {
  index: number
  message: Message
  lastIndex: boolean
  loading: boolean
  imageSrc: ImageFile[]
  modelName: string
  handleImageDelete: (e: any) => void
}

const ChatMessage: React.FC<ChatMessageProps> = ({
  index,
  message,
  imageSrc,
  lastIndex,
  modelName,
  loading,
  handleImageDelete,
}) => {
  const convertNewlinesToBreaks = (text: string) => {
    return text.replace(/\n/g, '<br>')
  }

  return (
    <>
      {index !== 0 && (
        <article
          className="flex px-3 py-2 text-black"
          aria-label="user-message"
        >
          <div className="w-16 flex flex-col justify-start items-center ai-answer">
            <Image
              src="/user.png"
              alt="User avatar"
              width="30"
              height="30"
              className="h-8 w-8 mr-4 rounded-sm"
              priority
            />
          </div>
          <div className="w-11/12 user-question">
            <div
              className="space-wrap"
              dangerouslySetInnerHTML={{
                __html: convertNewlinesToBreaks(message.question),
              }}
            />
            {
              <ImageListWithModal
                imageSrc={imageSrc}
                handleImageDelete={handleImageDelete}
              />
            }
          </div>
        </article>
      )}
      {
        <article
          className="flex px-3 py-2 bg-slate-100 color-black ease-in duration-300"
          aria-labelledby="ai-response"
        >
          <div className="w-16 flex flex-col justify-start items-center ai-answer">
            <Image
              src="/bot.png"
              alt="AI bot avatar"
              width="30"
              height="30"
              className={`h-8 w-8 mr-4 rounded-sm ${loading && lastIndex && 'animate-pulse'}`}
              priority
            />
            <label className="text-xs">{modelName}</label>
          </div>
          <div
            className=" bot-response answer space-wrap"
            dangerouslySetInnerHTML={{
              __html: convertNewlinesToBreaks(message.answer),
            }}
          />
        </article>
      }
    </>
  )
}

export default ChatMessage
