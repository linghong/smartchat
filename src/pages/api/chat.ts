import type { NextApiRequest, NextApiResponse } from 'next';
import { marked } from 'marked';

import { fetchDataFromPinecone } from '@/src/services/fetchDataFromPinecone';
import getGeminiChatCompletion from '@/src/services/llm/gemini';
import getClaudeChatCompletion from '@/src/services/llm/claude';
import { getGroqChatCompletion } from '@/src/services/llm/groq';
import {
  createEmbedding,
  getOpenAIChatCompletion
} from '@/src/services/llm/openai';
import getOpenModelChatCompletion from '@/src/services/llm/opensourceai';
import {
  extractMessageContent,
  extractSubjectTitle
} from '@/src/utils/chatMessageHelper';
import { ImageFile } from '@/src/types/chat';

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  const {
    basePrompt,
    question,
    namespace,
    selectedModel,
    chatHistory,
    imageSrc
  } = req.body;

  if (!question && imageSrc.length === 0) {
    return res.status(400).json({ message: 'No question in the request' });
  }

  if (!selectedModel?.value) {
    return res.status(500).json('Model name is missing.');
  }

  const base64ImageSrc = imageSrc?.map((imgSrc: ImageFile) => ({
    ...imgSrc,
    base64Image: imgSrc.base64Image.split(',')[1]
  }));

  // replacing newlines with spaces
  const sanitizedQuestion = question.trim().replaceAll('\n', ' ');

  try {
    let fetchedText = '';

    //fetch text from vector database
    if (namespace && namespace !== 'none') {
      const embeddedQuery = await createEmbedding(question);
      fetchedText = await fetchDataFromPinecone(embeddedQuery, namespace);
    }

    // get response from AI
    const { category } = selectedModel;
    let chatResponse: string | undefined | null;

    switch (category) {
      case 'openai':
        chatResponse = await getOpenAIChatCompletion(
          basePrompt,
          chatHistory,
          sanitizedQuestion,
          fetchedText,
          selectedModel,
          base64ImageSrc
        );
        break;

      case 'groq':
        chatResponse = await getGroqChatCompletion(
          basePrompt,
          chatHistory,
          sanitizedQuestion,
          fetchedText,
          selectedModel
        );
        break;

      case 'google':
        chatResponse = await getGeminiChatCompletion(
          basePrompt,
          chatHistory,
          sanitizedQuestion,
          fetchedText,
          selectedModel,
          base64ImageSrc
        );
        break;

      case 'anthropic':
        chatResponse = await getClaudeChatCompletion(
          basePrompt,
          chatHistory,
          sanitizedQuestion,
          fetchedText,
          selectedModel,
          base64ImageSrc
        );
        break;

      case 'hf-small':
      case 'hf-large':
        const baseUrl =
          category === 'hf-small'
            ? process.env.NEXT_PUBLIC_SERVER_URL
            : process.env.NEXT_PUBLIC_SERVER_GPU_URL;
        if (!baseUrl) {
          return res
            .status(500)
            .json(
              `Url address for posting the data to ${selectedModel} is missing`
            );
        }
        const url = `${baseUrl}/api/chat_${category === 'hf-small' ? 'cpu' : 'gpu'}`;
        chatResponse = await getOpenModelChatCompletion(
          basePrompt,
          chatHistory,
          sanitizedQuestion,
          fetchedText,
          selectedModel,
          url
        );
        break;

      default:
        return res.status(500).json('Invalid model category');
    }

    let answer: string;
    let subject: string;
    if (!chatResponse) {
      answer = "Sorry, I'm having trouble finding an answer to your question.";
      subject = 'Unknown';
    } else {
      answer = extractMessageContent(chatResponse);
      subject = extractSubjectTitle(chatResponse);
    }

    const htmlAnswer = marked(answer);

    res.status(200).json({ answer: htmlAnswer, subject });
  } catch (error: any) {
    console.error('An error occurred: ', error);
    res.status(500).json({ error: error.message || 'Something went wrong' });
  }
}
