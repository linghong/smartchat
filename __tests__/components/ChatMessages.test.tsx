import React from 'react';
import { render, screen, act } from '@testing-library/react';
import '@testing-library/jest-dom';

import ChatMessage from '@/src/components/ChatMessage';
import { Message, ImageFile } from '@/src/types/chat';

// Mock the console methods to suppress output during tests
beforeAll(() => {
  jest.spyOn(console, 'log').mockImplementation(() => {});
  jest.spyOn(console, 'warn').mockImplementation(() => {});
  jest.spyOn(console, 'error').mockImplementation(() => {});
});

afterAll(() => {
  jest.restoreAllMocks();
});


describe('ChatMessage Component', () => {
  let index: number;
  let message: Message;
  let loading: boolean;
  let lastIndex: boolean;
  let imageSrc: ImageFile[];
  let modelName: string;
  let handleImageDelete: jest.Mock;

  beforeEach(() => {
    message = {
      question: 'What is AI?',
      answer: 'AI stands for Artificial Intelligence.'
    };
    loading = false;
    lastIndex = true;
    handleImageDelete = jest.fn();
    imageSrc = [
      {
        base64Image: '/path/to/image1.png',
        mimeType: 'image/png',
        size: 5000,
        name: 'image1'
      },
      {
        base64Image: 'path/to/image2.png',
        mimeType: 'image/png',
        size: 8000,
        name: 'image2'
      }
    ];
    modelName = 'gpt-4o';

    render(
      <ChatMessage
        index={index}
        message={message}
        lastIndex={lastIndex}
        loading={loading}
        imageSrc={imageSrc}
        modelName={modelName}
        handleImageDelete={handleImageDelete}
      />
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders ChatMessage component correctly with question and answer', () => {
    expect(screen.getByText('What is AI?')).toBeInTheDocument();
    expect(
      screen.getByText('AI stands for Artificial Intelligence.')
    ).toBeInTheDocument();
  });

  it('renders the user and AI avatars', () => {
    expect(screen.getByAltText('User avatar')).toBeInTheDocument();
    expect(screen.getByAltText('AI avatar')).toBeInTheDocument();
  });

  it('applies animate-pulse class when loading and is the last index', () => {
    act(() => {
      render(
        <ChatMessage
          index={index}
          message={message}
          lastIndex={true}
          loading={true}
          imageSrc={imageSrc}
          modelName={modelName}
          handleImageDelete={handleImageDelete}
        />
      );
    });
    const botImages = screen.getAllByAltText('AI avatar') as HTMLElement[];
    expect(botImages[botImages.length - 1]).toHaveClass('animate-pulse');
  });

  it('does not apply animate-pulse class when not loading', () => {
    act(() => {
      render(
        <ChatMessage
          index={index}
          message={message}
          lastIndex={true}
          loading={false}
          imageSrc={imageSrc}
          modelName={modelName}
          handleImageDelete={handleImageDelete}
        />
      );
    });

    const botImages = screen.getAllByAltText('AI avatar') as HTMLElement[];
    expect(botImages[botImages.length - 1]).not.toHaveClass('animate-pulse');
  });

  it('ChatMessage component snapshot', () => {
    const { asFragment } = render(
      <ChatMessage
        index={index}
        message={message}
        lastIndex={lastIndex}
        loading={loading}
        imageSrc={imageSrc}
        modelName={modelName}
        handleImageDelete={handleImageDelete}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
