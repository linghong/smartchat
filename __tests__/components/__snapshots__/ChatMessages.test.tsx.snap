// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatMessage Component ChatMessage component snapshot 1`] = `
<DocumentFragment>
  <article
    aria-label="user-message"
    class="flex px-3 py-2 text-black"
  >
    <div
      class="w-16 flex flex-col justify-start items-center ai-answer"
    >
      <img
        alt="User avatar"
        class="h-8 w-8 mr-4 rounded-sm"
        data-nimg="1"
        decoding="async"
        fetchpriority="high"
        height="30"
        src="/_next/image?url=%2Fuser.png&w=64&q=75"
        srcset="/_next/image?url=%2Fuser.png&w=32&q=75 1x, /_next/image?url=%2Fuser.png&w=64&q=75 2x"
        style="color: transparent;"
        width="30"
      />
    </div>
    <div
      class="w-11/12 user-question"
    >
      <div
        class="space-wrap"
      >
        What is AI?
      </div>
      <div
        class="flex flex-row justify-start w-full p-2"
      >
        <div
          class="w-20 h-full mx-2 relative"
        >
          <div
            aria-label="Click to view larger image 1"
            class="mr-2 p-1 bg-clip-border bg-slate-100 border-2 border-violet-100"
            role="button"
            tabindex="0"
            title="Uploaded image thumbnail 1"
          >
            <div
              class="bg-no-repeat bg-center bg-contain min-h-[60px] h-full w-full cursor-pointer"
              style="background-image: url(/path/to/image1.png);"
            />
          </div>
        </div>
        <div
          class="w-20 h-full mx-2 relative"
        >
          <div
            aria-label="Click to view larger image 2"
            class="mr-2 p-1 bg-clip-border bg-slate-100 border-2 border-violet-100"
            role="button"
            tabindex="0"
            title="Uploaded image thumbnail 2"
          >
            <div
              class="bg-no-repeat bg-center bg-contain min-h-[60px] h-full w-full cursor-pointer"
              style="background-image: url(path/to/image2.png);"
            />
          </div>
        </div>
      </div>
    </div>
  </article>
  <article
    aria-labelledby="ai-response"
    class="flex px-3 py-2 bg-slate-100 color-black ease-in duration-300"
  >
    <div
      class="w-16 flex flex-col justify-start items-center ai-answer"
    >
      <img
        alt="AI bot avatar"
        class="h-8 w-8 mr-4 rounded-sm false"
        data-nimg="1"
        decoding="async"
        fetchpriority="high"
        height="30"
        src="/_next/image?url=%2Fbot.png&w=64&q=75"
        srcset="/_next/image?url=%2Fbot.png&w=32&q=75 1x, /_next/image?url=%2Fbot.png&w=64&q=75 2x"
        style="color: transparent;"
        width="30"
      />
      <label
        class="text-xs"
      >
        gpt-4o
      </label>
    </div>
    <div
      class=" bot-response answer space-wrap"
    >
      AI stands for Artificial Intelligence.
    </div>
  </article>
</DocumentFragment>
`;
